content/knowledge-base/guides/write-a-composition-function-in-python.md

```diff
@@ -0,0 +1,749 @@
+        region = req.observed.composite.resource["spec"]["region"]
+        names = req.observed.composite.resource["spec"]["names"]
+
+        # Add a desired S3 bucket for each name.
+        for name in names:
+            # Crossplane represents desired composed resources using a protobuf
+            # map of messages. This works a little like a Python defaultdict.
+            # Instead of assigning to a new key in the dict-like map, you access
+            # the key and mutate its value as if it did exist.
+            #
+            # The below code works because accessing the xbuckets-{name} key
+            # automatically creates a new, empty fnv1beta1.Resource message. The
+            # Resource message has a resource field containing an empty Struct
+            # object that can be populated from a dictionary by calling update.
+            #
+            # https://protobuf.dev/reference/python/python-generated/#map-fields
+            rsp.desired.resources[f"xbuckets-{name}"].resource.update(
+                {
+                    "apiVersion": "s3.aws.upbound.io/v1beta1",
+                    "kind": "Bucket",
+                    "metadata": {
+                        "annotations": {
+                            "crossplane.io/external-name": name,
+                        },
+                    },
+                    "spec": {
+                        "forProvider": {
+                            "region": region,
+                        },
+                    },
+                }
+            )
+
+        # Log what the function did. This will only appear in the function's pod
+        # logs. A function can use response.normal() and response.warning() to
+        # emit Kubernetes events associated with the XR it's operating on.
+        log.info("Added desired buckets", region=region, count=len(names))
+
+        return rsp
+```
+{{</expand>}}
+
+This code:
+
+1. Gets the observed composite resource from the `RunFunctionRequest`.
+1. Gets the region and bucket names from the observed composite resource.
+1. Adds one desired S3 bucket for each bucket name.
+1. Returns the desired S3 buckets in a `RunFunctionResponse`.
+
+
+
+{{<hint "tip">}}
+Crossplane provides a
+[software development kit](https://github.com/crossplane/function-sdk-python)
+(SDK) for writing composition functions in Python. This function uses utilities
+from the SDK. Read the
+[documentation](https://crossplane.github.io/function-sdk-python) for the SDK.
+{{</hint>}}
+
+{{<hint "important">}}
+The Python SDK automatically generates the `RunFunctionRequest` and
+`RunFunctionResponse` Python objects from a
+[Protocol Buffers](https://protobuf.dev) schema. You can see the schema in the
+[Buf Schema Registry](https://buf.build/crossplane/crossplane/docs/main:apiextensions.fn.proto.v1beta1).
+
+The fields of the generated Python objects behave similarly to builtin Python
+types like dictionaries and lists. You should be aware that there are some
```

Comment:
```suggestion
The fields of the generated Python objects behave similar to builtin Python
types like dictionaries and lists. Be aware that there are some
```

Reply:
I think similarly is correct here based on a quick Googling.

Reply:
"similarly" is an adverb describing verbs. In this context we're talking about a noun (builtin python type) so the adjective "similar" should be used.